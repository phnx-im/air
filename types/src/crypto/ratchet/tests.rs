// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::sync::LazyLock;

use rand::SeedableRng;

use crate::{
    codec::PhnxCodec,
    messages::{
        client_ds::{QsQueueMessagePayload, QsQueueMessageType},
        EncryptedQsQueueMessage,
    },
    time::TimeStamp,
};

use super::*;

// Test that the ratchet works.
#[test]
fn test_ratchet() {
    let ratchet_secret = RatchetSecret::random().unwrap();
    let mut sender_ratchet = QueueRatchet::try_from(ratchet_secret.clone()).unwrap();
    let mut receiver_ratchtet = QueueRatchet::try_from(ratchet_secret).unwrap();
    let message = QsQueueMessagePayload {
        timestamp: TimeStamp::now(),
        message_type: QsQueueMessageType::MlsMessage,
        payload: vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    };
    let encrypted_message = sender_ratchet.encrypt(message.clone()).unwrap();
    let plaintext: QsQueueMessagePayload = receiver_ratchtet.decrypt(encrypted_message).unwrap();
    assert_eq!(plaintext, message);
}

static RATCHET: LazyLock<QueueRatchet<EncryptedQsQueueMessage, QsQueueMessagePayload>> =
    LazyLock::new(|| {
        let mut rng = rand_chacha::ChaCha20Rng::from_seed([42; 32]);
        let secret = RatchetSecret::random_with_rng(&mut rng).unwrap();
        secret.try_into().unwrap()
    });

#[test]
fn ratchet_secret_serde_stability_json() {
    insta::assert_json_snapshot!(&*RATCHET);
}

#[test]
fn ratchet_secret_serde_stability_cbor() {
    let bytes = PhnxCodec::to_vec(&*RATCHET).unwrap();
    insta::assert_binary_snapshot!(".cbor", bytes);
}
