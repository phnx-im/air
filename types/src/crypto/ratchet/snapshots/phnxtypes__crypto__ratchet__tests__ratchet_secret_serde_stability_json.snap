---
source: types/src/crypto/ratchet/tests.rs
expression: "&*RATCHET"
snapshot_kind: text
---
{
  "sequence_number": 0,
  "secret": {
    "key": {
      "secret": [
        152,
        25,
        31,
        70,
        229,
        131,
        2,
        22,
        68,
        84,
        54,
        151,
        136,
        3,
        105,
        122,
        94,
        58,
        182,
        27,
        30,
        137,
        81,
        212,
        254,
        154,
        230,
        123,
        171,
        97,
        74,
        95
      ]
    }
  },
  "key": {
    "key": {
      "secret": [
        74,
        56,
        89,
        181,
        167,
        124,
        19,
        57,
        122,
        40,
        120,
        59,
        56,
        173,
        111,
        215,
        122,
        210,
        102,
        143,
        71,
        2,
        149,
        13,
        217,
        68,
        161,
        31,
        82,
        43,
        131,
        164
      ]
    }
  },
  "_phantom": null
}
